emission/user_model_josh/utility_model.py

```diff
@@ -0,0 +1,619 @@
+RANDOM_RADIUS = .3  # 300 meters around center of campus; for randomization
+N_TOP_TRIPS = 3 # Number of top trips we return for the user to look at
+
+
+class UserBase:
+
+    """ 
+    Stores all the users and stores the population of areas  
+    Also keeps state on other useful things that we need to know, like caches
+    """
+
+    def __init__(self):
+        self.users = {}
+        self.crowd_areas = {}
+        self.last_info = {}  ## so we only call google maps if we change things
+        self.old_trips = None
+        self.geocode_cache = {} # fewer calls to google maps
+
+    def add_user(self, user):
+        self.users[user.name] = user
+
+    def get_user(self, user_name):
+        return self.users[user_name]
+
+    def add_crowd(self, area):
+        self.crowd_areas[area.name] = area
+
+    def get_crowd_info(self, area_name):
+        return self.crowd_areas[area_name]
+
+    def geocode_with_cache(self, place):
+        coder = geo.Geocoder()
+        if place in self.geocode_cache:
+            return self.geocode_cache[place]
+        else:
+            coded = coder.geocode(place)
+            self.geocode_cache[place] = coded
+            return coded
+
+
+the_base = UserBase()
+
+
+class CampusTrip:
+
+    def __init__(self, score_list, time_duration, points, source):
+        self.time = score_list[0] 
+        self.sweat = score_list[1]
+        self.beauty = score_list[2]
+        self.social = score_list[3]
+        self.tot_score = sum(score_list)
+        self.time_duration = time_duration / float(60)
+        self.points = points
+        self.source = source
+
+    def make_points(self):
+        to_return = ""
+        for p in self.points:
+            to_return += str(p[0])
+            to_return += ","
+            to_return += str(p[1])
+            to_return += ","
+        return to_return
+
+
+    def make_for_browser(self):
+        return '%s;%s;%s;%s;%s;%s' % (self.beauty, self.time, self.social, self.sweat, self.time_duration, self.make_points())
+
+    def make_jsn(self):
+        return json.dumps({"time" : self.time, "beauty" : self.beauty, "social" : self.social, "sweat" : self.sweat, "duration" : self.time_duration, "points" : self.points})
+        
+    def make_json(self):
+        return json.dumps(self.make_for_browser())
+
+    def __repr__(self):
+        return "total score : %f || source : %s || beauty : %f || sweat : %f || time : %f || social : %f" % (self.tot_score, self.source, self.beauty, self.sweat, self.time, self.social)
+
+    def __eq__(self, other):
+        return self.make_points() == other.make_points()
+
+
+class UserModel:
+
+    """ 
+    User Model class  
+    Can do lots of cool things
+    """ 
+
+    def __init__(self, name, has_bike=False):
+        self.name = name
+        self.utilities = emmc.Counter()
+        self.has_bike = has_bike
+        self.user_base = the_base
+
+        self.user_base.add_user(self)
```

Comment:
Now that you have removed the username, how do you account for the difference between one user making multiple calls and multiple users making calls? The count will vary depending on how they are counted, right?